index.js

```diff
@@ -107,34 +103,75 @@ async function init() {
 
 	updateStatus('Retrieving directory info…');
 
-	await validateInput(repo);
+	const repoMetadata = await validateInput(`${user}/${repo}`);
+
+	const files = await listContent.viaTreesApi({
+		user,
+		repository: repo,
+		ref,
+		directory: decodeURIComponent(dir),
+		token: localStorage.token,
+		getFullData: true
+	});
 
-	const files = await listContent.viaTreesApi(`${repo}#${ref}`, decodeURIComponent(dir), localStorage.token);
 	if (files.length === 0) {
 		updateStatus('No files to download');
 		return;
 	}
 
-	updateStatus(`Downloading (0/${files.length}) files…`, '\n• ' + files.join('\n• '));
+	updateStatus(`Downloading (0/${files.length}) files…`, '\n• ' + files.map(file => file.path).join('\n• '));
 
 	let downloaded = 0;
 	const zip = new JSZip();
 	const controller = new AbortController();
-	const download = async path => {
-		const response = await fetch(`https://raw.githubusercontent.com/${repo}/${ref}/${path}`, {
+
+	const getBlobFromResponse = repoMetadata.private ?
+		async response => {
+			const {content, encoding} = await response.json();
+
+			if (encoding !== 'base64') {
+				throw new Error(`Unknown encoding "${encoding}"`);
+			}
+
+			let decodingResponse;
+			try {
+				decodingResponse = await fetch(`data:application/octet-stream;base64,${content}`);
+			} catch {
+				throw new Error('Decoding failed');
+			}
+
+			if (!decodingResponse.ok) {
+				throw new Error('Decoding failed');
+			}
+
+			return decodingResponse.blob();
```

Comment:
Can you shorten this function to

```js
			const {content, encoding} = await response.json();
			const decodingResponse = await fetch(`data:application/octet-stream;${encoding},${content}`);
			return decodingResponse.blob();
```

I don't think we need to check each step of the way. **If** the content isn't compatible, it will fail at once.

Reply:
We can, kind of. Need to wrap the whole function in a try/catch block. The `throw new Error('Decoding failed')` is required somewhere to be able to show the correct error message, but that should not be a problem.

Reply:
The API guarantees base64 encoding, so `base64` can be hardcoded https://developer.github.com/v3/git/blobs/#get-a-blob

```js
const {content} = await response.json();
const decodingResponse = await fetch(`data:application/octet-stream;base64,${content}`);
return decodingResponse.blob();
```

Reply:
> The `throw new Error('Decoding failed')` is required somewhere

I don't think that's ever gonna be a problem. The real error will bubble up via the existing `throw errow` anyway, if any.

Reply:
> The API guarantees base64 encoding, so `base64` can be hardcoded

Great catch. I thought there may be more encodings because it's contained in the response.

> I don't think that's ever gonna be a problem. The real error will bubble up via the existing `throw error` anyway, if any.

It will. The problem though is that we still have no way to detect errors that result from requests being blocked by ad blockers. Therefore, any error that can not specifically be identified will show the "disable your ad blocker" message. Therefore we need the specific "decoding failed" error.

Reply:
Find out what a failed decoding would look like before catching and creating an additional error manually. There's no point in blindly trying to plug holes that might not exist (like `Unknown encoding`)

Reply:
True. When we trust in GitHub not delivering corrupt data (which we should), there is no way that this fails. I guess I'm a little bit overcautious. 😁